File: packages/clawdstrike-policy/src/plugins/loader.ts
import fs from 'node:fs';
import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { createRequire } from 'node:module';

import type { CustomGuardFactory } from '../custom-registry.js';
import { CustomGuardRegistry } from '../custom-registry.js';
import {
  parsePluginManifest,
  type PluginCapabilities,
  type PluginManifest,
  type PluginResourceLimits,
} from './manifest.js';

const DEFAULT_CURRENT_VERSION = '0.1.0';

export type PluginLoadResult = {
  root: string;
  manifest: PluginManifest;
  registered: string[];
  executionMode: PluginExecutionMode;
};

export type PluginResolveOptions = {
  fromDir?: string;
};

export type PluginExecutionMode = 'node' | 'wasm';

export interface PluginLoaderOptions extends PluginResolveOptions {
  trustedOnly?: boolean;
  allowWasmSandbox?: boolean;
  currentClawdstrikeVersion?: string;
  maxResources?: Partial<PluginResourceLimits>;
}

export interface PluginInspectResult {
  root: string;
  manifest: PluginManifest;
  executionMode: PluginExecutionMode;
}

export class PluginLoader {
  private readonly options: Required<
    Pick<PluginLoaderOptions, 'trustedOnly' | 'allowWasmSandbox' | 'currentClawdstrikeVersion'>
  > & {
    fromDir: string;
    maxResources: Partial<PluginResourceLimits>;
  };
  private readonly inspected = new Map<string, PluginInspectResult>();

  constructor(options: PluginLoaderOptions = {}) {
    this.options = {
      fromDir: options.fromDir ?? process.cwd(),
      trustedOnly: options.trustedOnly ?? true,
      allowWasmSandbox: options.allowWasmSandbox ?? false,
      currentClawdstrikeVersion: options.currentClawdstrikeVersion ?? DEFAULT_CURRENT_VERSION,
      maxResources: options.maxResources ?? {},
    };
  }

  async inspect(pluginRef: string): Promise<PluginInspectResult> {
    const cached = this.inspected.get(pluginRef);
    if (cached) {
      return cached;
    }

    const root = resolvePluginRoot(pluginRef, this.options.fromDir);
    const manifestPath = path.join(root, 'clawdstrike.plugin.json');
    if (!fs.existsSync(manifestPath)) {
      throw new Error(`missing clawdstrike.plugin.json in ${root}`);
    }

    const manifestRaw = JSON.parse(fs.readFileSync(manifestPath, 'utf8')) as unknown;
    const manifest = parsePluginManifest(manifestRaw);

    this.validateTrustPolicy(manifest);
    this.validateCompatibility(manifest);
    this.validateCapabilityPolicy(manifest.capabilities, manifest.name, manifest.trust.level);
    this.validateResourceLimits(manifest.resources, manifest.name);

    const result: PluginInspectResult = {
      root,
      manifest,
      executionMode: manifest.trust.sandbox,
    };
    this.inspected.set(pluginRef, result);
    return result;
  }

  async loadIntoRegistry(
    pluginRef: string,
    registry: CustomGuardRegistry,
  ): Promise<PluginLoadResult> {
    const inspected = await this.inspect(pluginRef);

    // WASM path is intentionally scaffolded and guarded until sandbox runtime lands.
    if (inspected.executionMode === 'wasm') {
      throw new Error(
        `WASM plugin loading scaffold is present but runtime is not implemented yet: ${inspected.manifest.name}`,
      );
    }

    const registered: string[] = [];
    for (const g of inspected.manifest.guards) {
      const entryPath = path.resolve(inspected.root, g.entrypoint);
      const mod = await import(pathToFileURL(entryPath).href);
      const factory = extractFactory(mod);
      if (factory.id !== g.name) {
        throw new Error(`plugin guard id mismatch: manifest=${g.name} entrypoint=${factory.id}`);
      }
      registry.register(factory);
      registered.push(factory.id);
    }

    return {
      root: inspected.root,
      manifest: inspected.manifest,
      registered,
      executionMode: inspected.executionMode,
    };
  }

  clearCache(): void {
    this.inspected.clear();
  }

  private validateTrustPolicy(manifest: PluginManifest): void {
    if (this.options.trustedOnly && manifest.trust.level !== 'trusted') {
      throw new Error(`refusing to load untrusted plugin: ${manifest.name}`);
    }
    if (manifest.trust.sandbox === 'wasm' && !this.options.allowWasmSandbox) {
      throw new Error(
        `refusing to load wasm-sandboxed plugin until WASM sandbox is enabled: ${manifest.name}`,
      );
    }
  }

  private validateCompatibility(manifest: PluginManifest): void {
    const compat = manifest.clawdstrike;
    if (!compat) {
      return;
    }

    const current = parseSemver(this.options.currentClawdstrikeVersion);
    if (!current) {
      return;
    }

    if (compat.minVersion) {
      const min = parseSemver(compat.minVersion);
      if (min && compareSemver(current, min) < 0) {
        throw new Error(
          `plugin ${manifest.name} requires clawdstrike >= ${compat.minVersion} (current ${this.options.currentClawdstrikeVersion})`,
        );
      }
    }

    if (compat.maxVersion && !satisfiesMaxVersion(current, compat.maxVersion)) {
      throw new Error(
        `plugin ${manifest.name} requires clawdstrike <= ${compat.maxVersion} (current ${this.options.currentClawdstrikeVersion})`,
      );
    }
  }

  private validateCapabilityPolicy(
    capabilities: PluginCapabilities,
    pluginName: string,
    trustLevel: 'trusted' | 'untrusted',
  ): void {
    // Capability policy stubs: enforce high-risk defaults before sandbox exists.
    if (trustLevel === 'untrusted') {
      if (capabilities.subprocess) {
        throw new Error(`untrusted plugin ${pluginName} cannot request subprocess capability`);
      }
      if (capabilities.filesystem.write) {
        throw new Error(`untrusted plugin ${pluginName} cannot request filesystem write capability`);
      }
      if (capabilities.secrets.access) {
        throw new Error(`untrusted plugin ${pluginName} cannot request secrets access capability`);
      }
    }
  }

  private validateResourceLimits(resources: PluginResourceLimits, pluginName: string): void {
    const max = this.options.maxResources;
    if (typeof max.maxMemoryMb === 'number' && resources.maxMemoryMb > max.maxMemoryMb) {
      throw new Error(
        `plugin ${pluginName} maxMemoryMb=${resources.maxMemoryMb} exceeds loader limit ${max.maxMemoryMb}`,
      );
    }
    if (typeof max.maxCpuMs === 'number' && resources.maxCpuMs > max.maxCpuMs) {
      throw new Error(
        `plugin ${pluginName} maxCpuMs=${resources.maxCpuMs} exceeds loader limit ${max.maxCpuMs}`,
      );
    }
    if (typeof max.maxTimeoutMs === 'number' && resources.maxTimeoutMs > max.maxTimeoutMs) {
      throw new Error(
        `plugin ${pluginName} maxTimeoutMs=${resources.maxTimeoutMs} exceeds loader limit ${max.maxTimeoutMs}`,
      );
    }
  }
}

export async function loadTrustedPluginIntoRegistry(
  pluginRef: string,
  registry: CustomGuardRegistry,
  options: PluginLoaderOptions = {},
): Promise<PluginLoadResult> {
  const loader = new PluginLoader({ ...options, trustedOnly: true, allowWasmSandbox: false });
  return loader.loadIntoRegistry(pluginRef, registry);
}

export async function inspectPlugin(
  pluginRef: string,
  options: PluginLoaderOptions = {},
): Promise<PluginInspectResult> {
  const loader = new PluginLoader(options);
  return loader.inspect(pluginRef);
}

export function resolvePluginRoot(pluginRef: string, fromDir: string): string {
  const maybePath = path.isAbsolute(pluginRef) ? pluginRef : path.resolve(fromDir, pluginRef);
  if (fs.existsSync(maybePath)) {
    const stat = fs.statSync(maybePath);
    return stat.isDirectory() ? maybePath : path.dirname(maybePath);
  }

  const require = createRequire(import.meta.url);
  const pkgJsonPath = require.resolve(`${pluginRef}/package.json`, { paths: [fromDir] });
  return path.dirname(pkgJsonPath);
}

function extractFactory(mod: any): CustomGuardFactory {
  const candidate = mod?.factory ?? mod?.default ?? mod;
  if (!isFactory(candidate)) {
    throw new Error('invalid plugin guard entrypoint: expected CustomGuardFactory export');
  }
  return candidate;
}

function isFactory(value: any): value is CustomGuardFactory {
  return Boolean(value) && typeof value === 'object' && typeof value.id === 'string' && typeof value.build === 'function';
}

type Semver = [number, number, number];

function parseSemver(value: string): Semver | null {
  const m = /^(\d+)\.(\d+)\.(\d+)$/.exec(value);
  if (!m) {
    return null;
  }
  return [Number(m[1]), Number(m[2]), Number(m[3])];
}

function compareSemver(a: Semver, b: Semver): number {
  if (a[0] !== b[0]) return a[0] - b[0];
  if (a[1] !== b[1]) return a[1] - b[1];
  return a[2] - b[2];
}

function satisfiesMaxVersion(current: Semver, maxVersion: string): boolean {
  const maxStrict = parseSemver(maxVersion);
  if (maxStrict) {
    return compareSemver(current, maxStrict) <= 0;
  }

  const majorWildcard = /^(\d+)\.x$/.exec(maxVersion);
  if (majorWildcard) {
    return current[0] === Number(majorWildcard[1]);
  }

  const minorWildcard = /^(\d+)\.(\d+)\.x$/.exec(maxVersion);
  if (minorWildcard) {
    return current[0] === Number(minorWildcard[1]) && current[1] === Number(minorWildcard[2]);
  }

  // Unknown max format is treated as unconstrained at this scaffold phase.
  return true;
}

File: crates/clawdstrike/src/plugins/loader.rs
use std::path::{Path, PathBuf};

use crate::error::{Error, Result};

use super::manifest::{
    parse_plugin_manifest_toml, PluginManifest, PluginResourceLimits, PluginSandbox,
    PluginTrustLevel,
};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PluginExecutionMode {
    Native,
    Wasm,
}

#[derive(Clone, Debug)]
pub struct PluginInspectResult {
    pub root: PathBuf,
    pub manifest_path: PathBuf,
    pub manifest: PluginManifest,
    pub execution_mode: PluginExecutionMode,
}

#[derive(Clone, Debug)]
pub struct PluginLoadPlan {
    pub inspect: PluginInspectResult,
    pub guard_ids: Vec<String>,
}

#[derive(Clone, Debug)]
pub struct PluginLoaderOptions {
    pub from_dir: PathBuf,
    pub trusted_only: bool,
    pub allow_wasm_sandbox: bool,
    pub current_clawdstrike_version: String,
    pub max_resources: Option<PluginResourceLimits>,
}

impl Default for PluginLoaderOptions {
    fn default() -> Self {
        Self {
            from_dir: std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")),
            trusted_only: true,
            allow_wasm_sandbox: false,
            current_clawdstrike_version: "0.1.0".to_string(),
            max_resources: None,
        }
    }
}

#[derive(Clone, Debug)]
pub struct PluginLoader {
    options: PluginLoaderOptions,
}

impl PluginLoader {
    pub fn new(options: PluginLoaderOptions) -> Self {
        Self { options }
    }

    pub fn inspect(&self, plugin_ref: &str) -> Result<PluginInspectResult> {
        let root = resolve_plugin_root(plugin_ref, &self.options.from_dir)?;
        let manifest_path = root.join("clawdstrike.plugin.toml");
        let content = std::fs::read_to_string(&manifest_path).map_err(|e| {
            Error::ConfigError(format!(
                "failed to read plugin manifest {}: {}",
                manifest_path.display(),
                e
            ))
        })?;

        let manifest = parse_plugin_manifest_toml(&content)?;
        self.validate_trust_policy(&manifest)?;
        self.validate_compatibility(&manifest)?;
        self.validate_resource_limits(&manifest)?;

        Ok(PluginInspectResult {
            root,
            manifest_path,
            execution_mode: to_execution_mode(manifest.trust.sandbox),
            manifest,
        })
    }

    pub fn plan_load(&self, plugin_ref: &str) -> Result<PluginLoadPlan> {
        let inspect = self.inspect(plugin_ref)?;

        if inspect.execution_mode == PluginExecutionMode::Wasm {
            return Err(Error::ConfigError(format!(
                "WASM plugin loading scaffold is present but runtime is not implemented yet: {}",
                inspect.manifest.plugin.name
            )));
        }

        let guard_ids = inspect
            .manifest
            .guards
            .iter()
            .map(|g| g.name.clone())
            .collect::<Vec<_>>();

        Ok(PluginLoadPlan { inspect, guard_ids })
    }

    fn validate_trust_policy(&self, manifest: &PluginManifest) -> Result<()> {
        if self.options.trusted_only && manifest.trust.level != PluginTrustLevel::Trusted {
            return Err(Error::ConfigError(format!(
                "refusing to load untrusted plugin: {}",
                manifest.plugin.name
            )));
        }

        if manifest.trust.sandbox == PluginSandbox::Wasm && !self.options.allow_wasm_sandbox {
            return Err(Error::ConfigError(format!(
                "refusing to load wasm-sandboxed plugin until WASM sandbox is enabled: {}",
                manifest.plugin.name
            )));
        }

        Ok(())
    }

    fn validate_compatibility(&self, manifest: &PluginManifest) -> Result<()> {
        let Some(compat) = &manifest.clawdstrike else {
            return Ok(());
        };

        let Some(current) = parse_semver(&self.options.current_clawdstrike_version) else {
            return Ok(());
        };

        if let Some(min) = compat.min_version.as_deref().and_then(parse_semver) {
            if compare_semver(current, min).is_lt() {
                return Err(Error::ConfigError(format!(
                    "plugin {} requires clawdstrike >= {} (current {})",
                    manifest.plugin.name,
                    compat.min_version.as_deref().unwrap_or(""),
                    self.options.current_clawdstrike_version
                )));
            }
        }

        if let Some(max) = compat.max_version.as_deref() {
            if !satisfies_max_version(current, max) {
                return Err(Error::ConfigError(format!(
                    "plugin {} requires clawdstrike <= {} (current {})",
                    manifest.plugin.name, max, self.options.current_clawdstrike_version
                )));
            }
        }

        Ok(())
    }

    fn validate_resource_limits(&self, manifest: &PluginManifest) -> Result<()> {
        let Some(max) = &self.options.max_resources else {
            return Ok(());
        };

        if manifest.resources.max_memory_mb > max.max_memory_mb {
            return Err(Error::ConfigError(format!(
                "plugin {} max_memory_mb={} exceeds loader limit {}",
                manifest.plugin.name, manifest.resources.max_memory_mb, max.max_memory_mb
            )));
        }
        if manifest.resources.max_cpu_ms > max.max_cpu_ms {
            return Err(Error::ConfigError(format!(
                "plugin {} max_cpu_ms={} exceeds loader limit {}",
                manifest.plugin.name, manifest.resources.max_cpu_ms, max.max_cpu_ms
            )));
        }
        if manifest.resources.max_timeout_ms > max.max_timeout_ms {
            return Err(Error::ConfigError(format!(
                "plugin {} max_timeout_ms={} exceeds loader limit {}",
                manifest.plugin.name, manifest.resources.max_timeout_ms, max.max_timeout_ms
            )));
        }

        Ok(())
    }
}

pub fn resolve_plugin_root(plugin_ref: &str, from_dir: &Path) -> Result<PathBuf> {
    let raw = PathBuf::from(plugin_ref);
    let candidate = if raw.is_absolute() {
        raw
    } else {
        from_dir.join(raw)
    };

    if candidate.exists() {
        let meta = std::fs::metadata(&candidate).map_err(|e| {
            Error::ConfigError(format!(
                "failed to stat plugin reference {}: {}",
                candidate.display(),
                e
            ))
        })?;
        return Ok(if meta.is_dir() {
            candidate
        } else {
            candidate
                .parent()
                .map(Path::to_path_buf)
                .unwrap_or_else(|| PathBuf::from("."))
        });
    }

    Err(Error::ConfigError(format!(
        "plugin reference not found (only local paths are supported in scaffold mode): {}",
        plugin_ref
    )))
}

fn to_execution_mode(sandbox: PluginSandbox) -> PluginExecutionMode {
    match sandbox {
        PluginSandbox::Native => PluginExecutionMode::Native,
        PluginSandbox::Wasm => PluginExecutionMode::Wasm,
    }
}

type Semver = [u32; 3];

fn parse_semver(value: &str) -> Option<Semver> {
    let mut parts = value.split('.');
    let major = parts.next()?.parse::<u32>().ok()?;
    let minor = parts.next()?.parse::<u32>().ok()?;
    let patch = parts.next()?.parse::<u32>().ok()?;
    if parts.next().is_some() {
        return None;
    }
    Some([major, minor, patch])
}

fn compare_semver(a: Semver, b: Semver) -> std::cmp::Ordering {
    a.cmp(&b)
}

fn satisfies_max_version(current: Semver, max_version: &str) -> bool {
    if let Some(max) = parse_semver(max_version) {
        return compare_semver(current, max).is_le();
    }

    if let Some(major) = parse_major_wildcard(max_version) {
        return current[0] == major;
    }

    if let Some((major, minor)) = parse_minor_wildcard(max_version) {
        return current[0] == major && current[1] == minor;
    }

    true
}

fn parse_major_wildcard(value: &str) -> Option<u32> {
    let mut parts = value.split('.');
    let major = parts.next()?.parse::<u32>().ok()?;
    let wildcard = parts.next()?;
    if wildcard != "x" || parts.next().is_some() {
        return None;
    }
    Some(major)
}

fn parse_minor_wildcard(value: &str) -> Option<(u32, u32)> {
    let mut parts = value.split('.');
    let major = parts.next()?.parse::<u32>().ok()?;
    let minor = parts.next()?.parse::<u32>().ok()?;
    let wildcard = parts.next()?;
    if wildcard != "x" || parts.next().is_some() {
        return None;
    }
    Some((major, minor))
}

#[cfg(test)]
mod tests {
    use tempfile::TempDir;

    use super::*;

    fn write_manifest(dir: &TempDir, content: &str) {
        std::fs::write(dir.path().join("clawdstrike.plugin.toml"), content)
            .expect("write manifest");
    }

    #[test]
    fn inspect_trusted_native_manifest() {
        let dir = TempDir::new().expect("tempdir");
        write_manifest(
            &dir,
            r#"
[plugin]
version = "1.0.0"
name = "acme-plugin"

[[guards]]
name = "acme.guard"

[trust]
level = "trusted"
sandbox = "native"
"#,
        );

        let loader = PluginLoader::new(PluginLoaderOptions {
            from_dir: dir.path().to_path_buf(),
            ..PluginLoaderOptions::default()
        });

        let inspected = loader.inspect(".").expect("inspect");
        assert_eq!(inspected.execution_mode, PluginExecutionMode::Native);
        assert_eq!(inspected.manifest.plugin.name, "acme-plugin");
    }

    #[test]
    fn trusted_only_loader_rejects_untrusted() {
        let dir = TempDir::new().expect("tempdir");
        write_manifest(
            &dir,
            r#"
[plugin]
version = "1.0.0"
name = "acme-plugin"

[[guards]]
name = "acme.guard"

[trust]
level = "untrusted"
sandbox = "wasm"
"#,
        );

        let loader = PluginLoader::new(PluginLoaderOptions {
            from_dir: dir.path().to_path_buf(),
            ..PluginLoaderOptions::default()
        });

        let err = loader.inspect(".").expect_err("should reject untrusted");
        assert!(err
            .to_string()
            .contains("refusing to load untrusted plugin"));
    }

    #[test]
    fn loader_checks_min_version_compatibility() {
        let dir = TempDir::new().expect("tempdir");
        write_manifest(
            &dir,
            r#"
[plugin]
version = "1.0.0"
name = "acme-plugin"

[clawdstrike]
min_version = "9.9.9"

[[guards]]
name = "acme.guard"

[trust]
level = "trusted"
sandbox = "native"
"#,
        );

        let loader = PluginLoader::new(PluginLoaderOptions {
            from_dir: dir.path().to_path_buf(),
            current_clawdstrike_version: "0.1.0".to_string(),
            ..PluginLoaderOptions::default()
        });

        let err = loader
            .inspect(".")
            .expect_err("should reject incompatible version");
        assert!(err.to_string().contains("requires clawdstrike >="));
    }
}
